### PR TITLE
[js/webgpu] Optimize MatMul with M = 1

### PR DESCRIPTION
**Pull request by qjia7**
### Description

BUG #22031

In the demucs model, there are lots of MatMul ops with shapes like below:
`input[0]: [3448,1,512] | float32, input[1]: [512,1536] | float32, output[0]: [3448,1,1536] | float32`

We can see that for this kind of shape, the batch size is a big value, but M = 1. Our current algorithm is based on [M, N] to partition tiles, which is not efficient for such kind of shapes. This PR reshapes the inputs to improve the matmul performance.
Before:  [3448,1,512] x [512,1536] =  [3448,1,1536]
After: [1, 3448, 512] x [512, 1536] = [1, 3448, 1536] , then the output can be reshaped to [3448, 1, 1536]

The overall MatMul time in demucs model becomes 1778.45 ms from 4418.17 ms on my iGPUs.

### CODE REVIEW

**Comment by qjia7** [comment]
@guschmue @fs-eire Please take a look, thanks.

**Review by fs-eire — APPROVED** [review]

File: js/web/lib/wasm/jsep/webgpu/ops/matmul.ts [hunk 1]
```diff
@@ -201,6 +201,19 @@ export const matMul = (context: ComputeContext): void => {
   if (N < 8 && K < 8) {
     context.compute(createNaiveMatmulProgramInfo(context.inputs, { activation: '' }, outputShape));
   } else {
-    context.compute(createMatmulProgramInfo(context.inputs, { activation: '' }, outputShape));
+    const M = outputShape[outputShape.length - 2];
+    const batchA = ShapeUtil.size(context.inputs[0].dims.slice(0, -2));
+    const batchB = ShapeUtil.size(context.inputs[1].dims.slice(0, -2));
+    if (batchA !== 1 && M === 1 && batchB === 1) {
+      const reshapedA = context.inputs[0].reshape([1, batchA, K]);
+      const reshapedB = context.inputs[0].reshape([1, K, N]);
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/matmul.ts:209]: inputs[1]

**Review by guschmue — APPROVED** [review]

**Comment by guschmue** [comment]
CI not happy:
image

File: js/web/lib/wasm/jsep/webgpu/ops/matmul.ts (hunk 1 above)
Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/ops/matmul.ts:209]:
```suggestion
      const reshapedB = context.inputs[1].reshape([1, K, N]);
```

**Comment by qjia7** [comment]
> CI not happy: image

Done. Thanks @fs-eire fixing it.
My previous perf data was based on the first commit which hasn't reshaped B so that I didn't find it earlier.

**Review by guschmue — APPROVED** [review]

**Pull request merged**